### ISSUE TITLE
False negative/positive: Culture rules (MA0011, MA0075, MA0076, MA0185) ignore polymorphic types implementing IFormattable at runtime

### ISSUE THREAD

**Issue by dlyz** [meziantou/Meziantou.Analyzer#1262]
### Version of the Meziantou.Analyzer NuGet package

3.0.139

### Rule Identifier

MA0011, MA0075, MA0076, MA0185

### Target Framework

net10

### C# Language version

C# 14

### Description

There is an inaccuracy in how the culture-sensitive formatting rules (MA0011, MA0075, MA0076) and the string simplification rule (MA0185) evaluate types. These rules use related but incomplete type analysis and do not fully account for .NET runtime behavior regarding polymorphism.

When interpolating or stringifying values typed broadly as object or an unsealed class, MA0185 flags a false positive, advising the user to strip away string.Create(CultureInfo.InvariantCulture, ...). However, .NET's runtime string formatting engine dynamically checks whether an instance implements IFormattable. If the boxed runtime type is culture-sensitive (e.g., double, DateTime), the output will unexpectedly change based on the executing environment's regional settings.

Conversely, MA0075 and MA0076 exhibit false negatives by ignoring unconstrained types or polymorphic variables that might implement IFormattable at runtime. MA0011 has a related gap for explicit polymorphic ToString() calls, although its existing analysis is based on the availability of an IFormatProvider overload. While native CA1305 analysis also lacks polymorphic evaluation, Meziantou.Analyzer can improve upon it by capturing these subtle localization gaps.

Furthermore, the documentation across these rules lacks cohesion and feels fragmented. Even though MA0011, MA0075, MA0076, and MA0185 rely on related culture-sensitivity analysis, they describe it differently and often incompletely (such as omitting any mention of the underlying IFormattable logic). Individual pages tend to focus on specific example types or isolated special cases, which leaves the general rule of how culture-sensitive values are recognized unclear.

### Reproduction Steps

```csharp
// A culture-sensitive type packaged dynamically inside an object
object polymorphicValue = 42.5; // double implements IFormattable

// ❌ False Positive on MA0185:
// Removing the explicit culture can change the runtime formatting.
// The runtime detects IFormattable and uses its culture-aware ToString()
// implementation, so the result can depend on the current culture.
var brokenFix = string.Create(CultureInfo.InvariantCulture, $"{polymorphicValue}");

// ❌ False Negative on MA0076:
// The runtime value implements IFormattable, so its formatting can depend on culture.
string standardInterpolation = $"{polymorphicValue}";

// ❌ False Negative on MA0075:
// The implicit ToString() call does not use the IFormatProvider overload.
string implicitToString = "Value: " + polymorphicValue;

// ❌ False Negative on MA0011:
// The runtime value may have a culture-sensitive ToString() implementation.
// MA0107 reports this when explicitly enabled, but only for object type.
string explicitToString = polymorphicValue.ToString();
```

#### Note on Existing Rule MA0107

It is worth noting that Meziantou.Analyzer already includes MA0107 (Do not use object.ToString). However, this rule is strictly scoped to instances typed directly as object calling .ToString(), is disabled by default, and does not address the broader polymorphic hole across interface calls, non-sealed classes, or modern interpolation workflows handled by MA0011, MA0075, and MA0076.

#### Proposed Solution

> Note: This proposal was compiled based on behavior observations mentioned in this issue and an analysis of rules documentation. The underlying source code was not thoroughly audited, so additional edge cases may exist.

Classify types for which ToString is called implicitly or explicitly and then apply appropriate behavior. The algorithm is unified accross all related rules: MA0011, MA0075, MA0076, MA0185.

Suggested type classification:

- `CultureInvariant` - a value can be classified as culture-invariant when the analyzer can establish invariance from one of these sources:

  - **Sealed type that does not implement `IFormattable`/`ISpanFormattable`**.
  - **Known invariant types** such as `Guid`, `Version`, enums, etc.
  - **Known type + invariant format combination**, where the formatting contract guarantees culture-independent output, e.g. particular `DateTime`/`DateTimeOffset`/`TimeSpan` formats.
  - **Known type + invariant constant value**, where the analyzer already has a specific invariant-value rule, e.g. non-negative integer constants.
  - **`CultureInsensitiveTypeAttribute`**.
  - **Other existing analyzer-specific invariant cases** that are already recognized by the current implementation and should be preserved.

- `CultureSensitive` - types whose formatting can accept a culture/format provider and therefore could be culture-sensitive:

  - Types implementing `IFormattable`/`ISpanFormattable` that are not covered by the known types of `CultureInvariant`.

- `OpaqueRuntimeTypes` - compile-time types for which the runtime type is essentially opaque to the analyzer:

  - `object`
  - interfaces (except `IFormattable`/`ISpanFormattable` or derived from them)
  - generic type parameters (except constrained with a non-OpaqueRuntimeTypes, in that case the category is inherited from that constrain type).

- `UnsealedClasses` - non-sealed class types that are not known to be `CultureInvariant` or `CultureSensitive`: where derived runtime types can inherit from `IFormattable`. It is theoretically possible to detect SOME of the hierarchies that actually have derived `IFormattable` class, that will make the class `CultureSensitive` but this is not so easy and is not a full solution anyway.

Suggested behavior:

* **`CultureInvariant`**

  * No culture-sensitivity warnings.
  * Culture-removing/simplifying code fixes are allowed.

* **`CultureSensitive`**

  * Report culture-sensitivity warnings where applicable.
  * Do not suggest simplifications that remove the explicit culture/provider.

* **`OpaqueRuntimeTypes`**

  * Treat as culture-sensitive by default, since the runtime type may implement `IFormattable`.
  * Provide a dedicated configuration option to treat opaque runtime types as culture-invariant for projects that prefer fewer diagnostics.

* **`UnsealedClasses`**

  * Treat as culture-invariant by default. This avoids false positives for ordinary unsealed classes whose `ToString()` implementations are themselves culture-invariant.
  * Provide a dedicated configuration option to treat unsealed classes as culture-sensitive for projects that prefer a more conservative analysis.

### MA0011 and explicit `ToString()`

MA0011 is broader: it covers calls to methods for which a culture-aware overload accepting an `IFormatProvider` exists. MA0075 and MA0076, on the other hand, specifically address culture-sensitive `ToString()` conversions, whether explicit or implicit.

For polymorphic `ToString()` calls, there are several possible approaches:

* **Special-case `ToString()` in MA0011** — keeps the existing rule responsible for the diagnostic, but weakens its single-responsibility boundary.
* **Extend MA0107** — make the existing "Do not use object.ToString" rule cover polymorphic `ToString()` calls. This is conceptually clean, but its existing opt-in configuration means users who have not enabled MA0107 would not get the protection and the name has to be changed.
* **Introduce a new dedicated rule** — keep MA0011 and MA0107 semantics unchanged and provide a dedicated diagnostic for potentially culture-sensitive explicit `ToString()` calls. This avoids changing existing rule behavior/configuration, at the cost of another rule.

#### Documentation Suggestions

* Unify the Documentation: Because MA0011 (explicit ToString()), MA0075 (implicit ToString in concatenations), MA0076 (string interpolation), and MA0185 (string creation simplification) share the same underlying concept of culture-sensitivity detection, their documentation pages should be unified or cross-referenced with a shared explanation of this behavior.

* Modern Fix Recommendations: For MA0011 and MA0075, the documentation should suggest using modern interpolated strings via string.Create(CultureInfo.InvariantCulture, $"{value}") as a recommended alternative alongside the standard ToString(CultureInfo.InvariantCulture) fixes. For implicit conversions in MA0075, utilizing string interpolation factories often serves as the most elegant and clean resolution path.

#### Alternative Minimal Proposal

If updating the type resolution rules across all four analyzers is too major of a breaking change or too complex to implement, a simpler alternative would be to make MA0185 conservative when deciding whether string.Create can be removed.

Specifically, MA0185 should only suggest removing the explicit CultureInfo.InvariantCulture when the analyzer can establish that the expression is for sure culture-invariant. If the expression has an opaque runtime type such as object, an interface, or an unconstrained generic type, or otherwise cannot be proven culture-invariant, MA0185 should not suggest the simplification.

Even if fixing MA0011/MA0075/MA0076 is deferred, this would prevent MA0185 from recommending a semantics-changing refactoring that can make the output culture-dependent.

**Comment by meziantou** [comment]
I've made some progress on it. I now have 3 cases: culture-insensitive, culture-sensitive (e.g. IFormattable), maybe culture-sensitive (object, interface, unsealed class, etc.).

At the moment I'm not sure what to do with the "maybe culture-sensitive" case. Should it report by default, or just it be behind an option.

```c#
void Sample(object obj)
{
    _ = "sample" + obj; 
}
```

Should it report and suggest `string.Create(CultureInfo.InvariantCulture, $"sample{obj}")`. Or should it be behind an opt-in option? I think the option would be less annoying for users. Otherwise it may be very verbose and report lots of cases that are ok.

**Comment by dlyz** [comment]
> Should it report and suggest string.Create(CultureInfo.InvariantCulture, $"sample{obj}"). Or should it be behind an opt-in option? I think the option would be less annoying for users. Otherwise it may be very verbose and report lots of cases that are ok.

This is why I suggested having two categories for 'maybe culture-sensitive'. If the type is opaque (object, interface, generic), then the user doesn't really know the type and **should** assume that it could be culture-sensitive. For non-sealed classes, the situation is a little different, because I assume it is a very rare situation where a base class does not implement IFormattable but some derived ones do. I guess the non-sealed class case would be responsible for the most annoying false positives, but the opaque type case would not.

So it would be great to have two separate options for each of the cases and I think that the good default is - treat opaque as culture-sensitive, and non-sealed classes as culture-invariant.

**Comment by meziantou** [comment]
Thanks for the clarification! Your suggestion makes sense. I'll see how I can introduce the 4th category and the associated options
